dss/api/subscriptions.py

```diff
@@ -3,35 +3,174 @@
+                                 id=uuid)
+    if not id_exists:
+        raise DSSException(requests.codes.not_found, "not_found", "Cannot find subscription!")
 
+    response = es_client.get(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                             doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                             id=uuid)
+
+    source = response['_source']
+    source['uuid'] = uuid
+    source['replica'] = replica
+
+    if source['owner'] != owner:
+        raise DSSException(requests.codes.forbidden, "forbidden", "Your credentials can't access this subscription!")
+
+    return jsonify(source), requests.codes.okay
+
+
+@dss_handler
 def find(replica: str):
-    pass
+    owner = request.token_info['email']
+    es_client = ElasticsearchClient.get(logger)
+
+    search_obj = Search(using=es_client,
+                        index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                        doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE)
+    search = search_obj.query({'match': {'owner': owner}})
+
+    responses = [{
+        'uuid': hit.meta.id,
+        'replica': replica,
+        'owner': owner,
+        'callback_url': hit.callback_url,
+        'query': hit.query.to_dict()}
+        for hit in search.scan()]
 
+    full_response = {'subscriptions': responses}
+    return jsonify(full_response), requests.codes.okay
 
+
+@dss_handler
 def put(extras: dict, replica: str):
-    pass
+    uuid = str(uuid4())
+    query = extras['query']
+    owner = request.token_info['email']
+
+    es_client = ElasticsearchClient.get(logger)
+
+    index_mapping = {
+        "mappings": {
+            DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE: {
+                "properties": {
+                    "owner": {
+                        "type": "string",
+                        "index": "not_analyzed"
+                    }
+                }
+            }
+        }
+    }
+    get_elasticsearch_index(es_client, DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME, logger, index_mapping)
+
+    percolate_registration = _register_percolate(es_client, uuid, query)
+
+    if percolate_registration['created']:
+        logger.debug(f"Percolate query registration succeeded:\n{percolate_registration}")
+    else:
+        logger.critical(f"Percolate query registration failed:\n{percolate_registration}")
+        raise DSSException(requests.codes.internal_server_error,
+                           "internal_server_error",
+                           "Unable to register elasticsearch percolate query!")
+
+    extras['owner'] = owner
+    subscription_registration = _register_subscription(es_client, uuid, extras)
+    if subscription_registration['created']:
+        logger.debug(f"Event Subscription succeeded:\n{subscription_registration}")
+    else:
+        logger.critical(f"Event Subscription failed:\n{subscription_registration}")
+
+        # Delete percolate query to make sure queries and subscriptions are in sync.
```

Comment:
To help ensure the percolate query gets removed if there is a problem adding the subscription, it seems try/except would be a useful construct here.
You were clearly careful with your error checking and I think this would pretty reliable as it stands, yet try/except is worth considering.